### PR TITLE
Add compilation overview section

### PR DESCRIPTION
**Pull request by mhegazy**
Add the content from my reply in https://github.com/Microsoft/TypeScript/issues/5550#issuecomment-154573103 to the wiki.

### CODE REVIEW

**Comment by sebald** [comment]
As someone who has no real experience with compilers and the `tsc` especially, the added text really helps to get a grasp of the steps that are taken to create a `program` and compile/transpile the source code.

:+1:

File: Architectural-Overview.md [hunk 1]
```diff
@@ -37,6 +37,30 @@ The language service supports the common set of a typical editor operations like
...
 
+## Overview of the compilation process
+
+The process starts with preprocessing. the preprocessor figures out what files should be included in the compilation by following references (/// references and import statements).
+
+The parser then generates AST `Node`s. These are just abstract representation of the user input in a tree format. A `SourceFile` object represents an AST for a given file with some additional information like file name and source text. 
+
+The binder then passes over the AST nodes and generates and binds `Symbol`s; these are one per named entity, each `Symbol` can have multiple `Node`s, e.g.: multiple declarations of `namespace`, or `class` and a `namespace` with the same name, etc.. the binder also handles scopes and makes sure that each symbol is created in the correct enclosing function/block scope.
+
+Generating a `SourceFile` (along with its `Symbol`s) are done through calling the `createSourceFile` API.
```
Comment (sandersn) [inline Architectural-Overview.md]: Generating ... **is** done

File: Architectural-Overview.md [hunk 2]
```diff
@@ -37,6 +37,30 @@ The language service supports the common set of a typical editor operations like
...
+
+So far the AST/Symbols represent the Infomation of a single file. But `namespace`s for instance can span multiple files, so are `interface`s, `class`s, etc.. so the next step is build a global view of all files in the compilation by building a `Program`.
```
Comment (sandersn) [inline Architectural-Overview.md]: Infomation => information

Comment (sandersn) [inline Architectural-Overview.md]: "so are `interface`s, `class`s, etc .."

I would reword this as "and so can interfaces and classes". (getting rid of the code formatting makes it easier to use 'classes' as the plural)

File: Architectural-Overview.md [hunk 3]
```diff
@@ -37,6 +37,30 @@ The language service supports the common set of a typical editor operations like
...
+
+The parser then generates AST `Node`s. These are just abstract representation of the user input in a tree format. A `SourceFile` object represents an AST for a given file with some additional information like file name and source text. 
+
+The binder then passes over the AST nodes and generates and binds `Symbol`s; these are one per named entity, each `Symbol` can have multiple `Node`s, e.g.: multiple declarations of `namespace`, or `class` and a `namespace` with the same name, etc.. the binder also handles scopes and makes sure that each symbol is created in the correct enclosing function/block scope.
+
+Generating a `SourceFile` (along with its `Symbol`s) are done through calling the `createSourceFile` API.
+
+So far the AST/Symbols represent the Infomation of a single file. But `namespace`s for instance can span multiple files, so are `interface`s, `class`s, etc.. so the next step is build a global view of all files in the compilation by building a `Program`. 
+
+A `Program` is a collection of `SourceFile`s and a set of `CompilerOptions`. A `Program` is created by calling the `createProgram` API. A program can have only one file.
```
Comment (sandersn) [inline Architectural-Overview.md]: A program can only have one file? But you just said a `Program` is a collection of `SourceFile`s.

File: Architectural-Overview.md [hunk 4]
```diff
@@ -37,6 +37,30 @@ The language service supports the common set of a typical editor operations like
...
+
+Generating a `SourceFile` (along with its `Symbol`s) are done through calling the `createSourceFile` API.
+
+So far the AST/Symbols represent the Infomation of a single file. But `namespace`s for instance can span multiple files, so are `interface`s, `class`s, etc.. so the next step is build a global view of all files in the compilation by building a `Program`. 
+
+A `Program` is a collection of `SourceFile`s and a set of `CompilerOptions`. A `Program` is created by calling the `createProgram` API. A program can have only one file. 
+
+From a `Program` instance a `TypeCheker` can be created. `TypeChecker` is the core of the TypeScript type system. it is the part responsible to figuring out relationships between `Symbols` from different file, assigning `Type`s to `Symbol`s and generating any semantic `Diagnostic`s (i.e. errors).
```
Comment (sandersn) [inline Architectural-Overview.md]: typo in first use of `TypeCheker`.

File: Architectural-Overview.md [hunk 5]
```diff
@@ -37,6 +37,30 @@ The language service supports the common set of a typical editor operations like
...
+
+The first thing a `TypeCheker` will do is to consolidate all the `Symbol`s from different `SourceFile`s into a single view, and build a single Symbol Table by "merging" any common `Symbol`s (e.g. `namespace`s spanning multiple files).
```
Comment (sandersn) [inline Architectural-Overview.md]: same typo here

File: Architectural-Overview.md [hunk 6]
```diff
@@ -37,6 +37,30 @@ The language service supports the common set of a typical editor operations like
...
+
+A `Program` is a collection of `SourceFile`s and a set of `CompilerOptions`. A `Program` is created by calling the `createProgram` API. A program can have only one file. 
+
+From a `Program` instance a `TypeCheker` can be created. `TypeChecker` is the core of the TypeScript type system. it is the part responsible to figuring out relationships between `Symbols` from different file, assigning `Type`s to `Symbol`s and generating any semantic `Diagnostic`s (i.e. errors).
+
+The first thing a `TypeCheker` will do is to consolidate all the `Symbol`s from different `SourceFile`s into a single view, and build a single Symbol Table by "merging" any common `Symbol`s (e.g. `namespace`s spanning multiple files).  
+
+After initializing the original state, the `TypeCheker` is ready to answer any questions about the program, things like: what is the `Symbol` for this `Node`, or what is the `Type` of this `Symbol`, what `Symbol`s are visible at a location in the AST, or what are the `Signature`s for a function declaration, or what errors would be produced for a file.
```
Comment (sandersn) [inline Architectural-Overview.md]: same typo here

File: Architectural-Overview.md [hunk 7]
```diff
@@ -37,6 +37,30 @@ The language service supports the common set of a typical editor operations like
...
+
+The `TypeCheker` computes everything lazely; it only "resolves" the necessary information to answer a question, the checker will only examine Nodes/Symbols/Types that contribute to the question at hand and will not attempt to examine additional entities.
```
Comment (sandersn) [inline Architectural-Overview.md]: and here

Comment (sandersn) [inline Architectural-Overview.md]: Put a period after "... answer a question." Otherwise the sentence runs on for too long.

File: Architectural-Overview.md (hunk 3 above)
Comment (mhegazy) [inline Architectural-Overview.md]: i meant as few as one file. but now i see it is confusing. removing.

**Comment by mhegazy** [comment]
@sandersn any more comments?

File: Architectural-Overview.md [hunk 8]
```diff
@@ -37,6 +37,30 @@ The language service supports the common set of a typical editor operations like
 
 * **Signature:** There are three types of signatures in the language: call, construct and index signatures.
 
+## Overview of the compilation process
+
+The process starts with preprocessing. the preprocessor figures out what files should be included in the compilation by following references (/// references and import statements).
```
Comment (DanielRosenwasser) [inline Architectural-Overview.md]: Split into two lines, correct casing on first character of second sentence.

Write \`/// <reference ... />\` comments

File: Architectural-Overview.md [hunk 9]
```diff
@@ -37,6 +37,30 @@ The language service supports the common set of a typical editor operations like
...
+
+The parser then generates AST `Node`s. These are just abstract representation of the user input in a tree format. A `SourceFile` object represents an AST for a given file with some additional information like file name and source text.
```
Comment (DanielRosenwasser) [inline Architectural-Overview.md]: Split sentences

"just **an** abstract representation"

"like **the** file name and source text"

File: Architectural-Overview.md [hunk 10]
```diff
@@ -37,6 +37,30 @@ The language service supports the common set of a typical editor operations like
 
 * **Signature:** There are three types of signatures in the language: call, construct and index signatures.
 
+## Overview of the compilation process
+
+The process starts with preprocessing. the preprocessor figures out what files should be included in the compilation by following references (/// references and import statements).
+
+The parser then generates AST `Node`s. These are just abstract representation of the user input in a tree format. A `SourceFile` object represents an AST for a given file with some additional information like file name and source text. 
+
+The binder then passes over the AST nodes and generates and binds `Symbol`s; these are one per named entity, each `Symbol` can have multiple `Node`s, e.g.: multiple declarations of `namespace`, or `class` and a `namespace` with the same name, etc.. the binder also handles scopes and makes sure that each symbol is created in the correct enclosing function/block scope.
```
Comment (DanielRosenwasser) [inline Architectural-Overview.md]: The binder then passes over the AST nodes and generates and binds `Symbol`s.
One `Symbol` exists one per named entity.
There is a subtle distinction but several declaration nodes can name the same entity.
That means that sometimes different `Node`s will have the same `Symbol`, and each `Symbol` keeps track of its declaration `Node`s.
For example, a `class` and a `namespace` with the same name can *merge* and will have the same `Symbol`.
The binder also handles scopes and makes sure that each `Symbol` is created in the correct enclosing scope.

File: Architectural-Overview.md [hunk 11]
```diff
@@ -37,6 +37,30 @@ The language service supports the common set of a typical editor operations like
...
+
+Generating a `SourceFile` (along with its `Symbol`s) is done through calling the `createSourceFile` API.
+
+So far the AST/Symbols represent the infomation of a single file. But `namespace`s for instance can span multiple files, so can interfaces, classes, etc.. so the next step is build a global view of all files in the compilation by building a `Program`.
```
Comment (DanielRosenwasser) [inline Architectural-Overview.md]: Split sentences

So far, `Symbol`s represent named entities as seen within a single file, but several declarations can merge multiple files, so the next step is to build a global view of all files in the compilation  by building a `Program`.

File: Architectural-Overview.md [hunk 12]
```diff
@@ -37,6 +37,30 @@ The language service supports the common set of a typical editor operations like
...
+
+The parser then generates AST `Node`s. These are just abstract representation of the user input in a tree format. A `SourceFile` object represents an AST for a given file with some additional information like file name and source text. 
+
+The binder then passes over the AST nodes and generates and binds `Symbol`s; these are one per named entity, each `Symbol` can have multiple `Node`s, e.g.: multiple declarations of `namespace`, or `class` and a `namespace` with the same name, etc.. the binder also handles scopes and makes sure that each symbol is created in the correct enclosing function/block scope.
+
+Generating a `SourceFile` (along with its `Symbol`s) is done through calling the `createSourceFile` API.
+
+So far the AST/Symbols represent the infomation of a single file. But `namespace`s for instance can span multiple files, so can interfaces, classes, etc.. so the next step is build a global view of all files in the compilation by building a `Program`. 
+
+A `Program` is a collection of `SourceFile`s and a set of `CompilerOptions`. A `Program` is created by calling the `createProgram` API.
```
Comment (DanielRosenwasser) [inline Architectural-Overview.md]: Split sentences

File: Architectural-Overview.md [hunk 13]
```diff
@@ -37,6 +37,30 @@ The language service supports the common set of a typical editor operations like
...
+
+The binder then passes over the AST nodes and generates and binds `Symbol`s; these are one per named entity, each `Symbol` can have multiple `Node`s, e.g.: multiple declarations of `namespace`, or `class` and a `namespace` with the same name, etc.. the binder also handles scopes and makes sure that each symbol is created in the correct enclosing function/block scope.
+
+Generating a `SourceFile` (along with its `Symbol`s) is done through calling the `createSourceFile` API.
+
+So far the AST/Symbols represent the infomation of a single file. But `namespace`s for instance can span multiple files, so can interfaces, classes, etc.. so the next step is build a global view of all files in the compilation by building a `Program`. 
+
+A `Program` is a collection of `SourceFile`s and a set of `CompilerOptions`. A `Program` is created by calling the `createProgram` API. 
+
+From a `Program` instance a `TypeChecker` can be created. `TypeChecker` is the core of the TypeScript type system. it is the part responsible to figuring out relationships between `Symbols` from different file, assigning `Type`s to `Symbol`s and generating any semantic `Diagnostic`s (i.e. errors).
```
Comment (DanielRosenwasser) [inline Architectural-Overview.md]: Split sentences

it -> It

responsible to -> responsible for

from different file -> from different files

comma after "`Symbol`s.

File: Architectural-Overview.md [hunk 14]
```diff
@@ -37,6 +37,30 @@ The language service supports the common set of a typical editor operations like
...
+
+The first thing a `TypeChecker` will do is to consolidate all the `Symbol`s from different `SourceFile`s into a single view, and build a single Symbol Table by "merging" any common `Symbol`s (e.g. `namespace`s spanning multiple files).  
+
+After initializing the original state, the `TypeChecker` is ready to answer any questions about the program. Such questions can be: what is the `Symbol` for this `Node`, or what is the `Type` of this `Symbol`, what `Symbol`s are visible at a location in the AST, or what are the `Signature`s for a function declaration, or what errors would be produced for a file.
```
Comment (DanielRosenwasser) [inline Architectural-Overview.md]: Split sentences and turn this into list form:

> After initializing the original state, the `TypeChecker` is ready to answer any questions about the program.
 Such "questions" might be:

> * What is the `Symbol` for this `Node`?
* What is the `Type` of this `Symbol`?
* What `Symbol`s are visible in this portion of the AST?
* What are the available `Signature`s for a function declaration?
* What errors should be reported for a file?

File: Architectural-Overview.md [hunk 15]
```diff
@@ -37,6 +37,30 @@ The language service supports the common set of a typical editor operations like
...
+
+The `TypeChecker` computes everything lazely; it only "resolves" the necessary information to answer a question, the checker will only examine Nodes/Symbols/Types that contribute to the question at hand and will not attempt to examine additional entities.
```
Comment (DanielRosenwasser) [inline Architectural-Overview.md]: lazely -> lazily

Comma to period & split sentences

`Node`s, `Symbol`s, and `Type`s

File: Architectural-Overview.md [hunk 16]
```diff
@@ -37,6 +37,30 @@ The language service supports the common set of a typical editor operations like
...
+
+An `Emitter` can also be created from a given `Program`. The `Emitter` is responsible for generating the desired output for a given `SourceFile`; this includes `.js`, `.jsx`, `.d.ts`, and `.js.map` outputs.
```
Comment (DanielRosenwasser) [inline Architectural-Overview.md]: Split sentences

**Comment by sandersn** [comment]
:1+:
